modules/transaction-payment/src/lib.rs

```diff
@@ -1391,21 +1400,32 @@ impl<T: Config + Send + Sync> TransactionPayment<T::AccountId, PalletBalanceOf<T
 where
 	PalletBalanceOf<T>: Send + Sync + FixedPointOperand,
 {
-	fn reserve_fee(who: &T::AccountId, weight: Weight) -> Result<PalletBalanceOf<T>, DispatchError> {
-		let fee = Pallet::<T>::weight_to_fee(weight);
-		<T as Config>::Currency::reserve_named(&RESERVE_ID, who, fee)?;
+	fn reserve_fee(
+		who: &T::AccountId,
+		fee: PalletBalanceOf<T>,
+		named: Option<ReserveIdentifier>,
+	) -> Result<PalletBalanceOf<T>, DispatchError> {
+		let fee = Pallet::<T>::calculate_final_fee(fee, None);
+		Pallet::<T>::native_then_alternative_or_default(who, fee)?;
+		T::Currency::reserve_named(&named.unwrap_or(RESERVE_ID), who, fee)?;
 		Ok(fee)
 	}
 
-	fn unreserve_fee(who: &T::AccountId, fee: PalletBalanceOf<T>) {
-		<T as Config>::Currency::unreserve_named(&RESERVE_ID, who, fee);
+	fn unreserve_fee(
+		who: &T::AccountId,
+		fee: PalletBalanceOf<T>,
+		named: Option<ReserveIdentifier>,
+	) -> PalletBalanceOf<T> {
+		let fee = Pallet::<T>::calculate_final_fee(fee, None);
+		<T as Config>::Currency::unreserve_named(&named.unwrap_or(RESERVE_ID), who, fee)
 	}
 
 	fn unreserve_and_charge_fee(
 		who: &T::AccountId,
 		weight: Weight,
```

Comment:
this should be passing Balance as well? we will want to apply multiplier here

Reply:
still use weight, as `unreserve_and_charge_fee` is only used by module evm scheduled call. if change to Balance, then module evm either depend on payment module, or add an etra `weight_to_fee` method to trait which seems no benefit.